GPUCollection/LLVMBitCodeVisitor.cs

```diff
@@ -38,23 +51,27 @@ internal string WalkTree(Expression expression)
         /// <summary>
         /// Modifiying https://github.com/paulsmith/getting-started-llvm-c-api/blob/master/sum.c
         /// </summary>
-        private void InitializeLLVMWriting()
+        private void InitializeLLVMWriting(Expression expression)
         {
             functionNumber = 0;
             module = LLVM.ModuleCreateWithName(ModuleNameString);
-            LLVMTypeRef mainRetType = LLVM.FunctionType(LLVM.Int32Type(), new LLVMTypeRef[0], false);
-            LLVMValueRef mainFunc = LLVM.AddFunction(module, "main", mainRetType);
+            LLVMTypeRef mainRetType = LLVM.FunctionType(GetLLVMTypeRef(expression), new LLVMTypeRef[] { GetLLVMTypeRef(expression) }, false);
+            mainFunc = LLVM.AddFunction(module, "CSMain", mainRetType);
+
             mainBuilder = LLVM.CreateBuilder();
-            LLVMBasicBlockRef mainBlock = LLVM.AppendBasicBlock(mainFunc, "MainEntry");
+            LLVMBasicBlockRef mainBlock = LLVM.AppendBasicBlock(mainFunc, "EntryBlock");
             LLVM.PositionBuilderAtEnd(mainBuilder, mainBlock);
+
+            // Initialize the first statement as a constant
+            mainFunctionRefStack = new Stack<LLVMValueRef>();
         }
 
         /// <summary>
         /// Modifiying https://github.com/paulsmith/getting-started-llvm-c-api/blob/master/sum.c
         /// </summary>
         private string FinalizeLLVMWriting()
         {
-            LLVM.BuildRet(mainBuilder, lastLLVMFunctionCalledFromMain);
+            LLVM.BuildRet(mainBuilder, mainFunctionRefStack.Pop());
```

Comment:
For Vulkan and DirectX, you'd have to write to an output buffer instead of returning the value directly

Reply:
Right. I was trying to make sure I was understanding the expression tree correctly before writing the correct bitcode for the GPU's so I hadn't done this yet. Seems like a good next step 😄